athena-federation-sdk/src/main/java/com/amazonaws/athena/connector/lambda/data/BlockUtils.java

```diff
@@ -1031,6 +1053,271 @@ else if (value instanceof Boolean && (boolean) value) {
+        Field keyValueStructField;
+        if (children.size() != 1) {
+            throw new IllegalStateException("Invalid Arrow Map schema: " + field);
+        }
+        else {
+            keyValueStructField = children.get(0);
+            if (!ENTRIES.equals(keyValueStructField.getName()) || !(keyValueStructField.getType() instanceof ArrowType.Struct)) {
+                throw new IllegalStateException("Invalid Arrow Map schema: " + field);
+            }
+        }
+
+        List<Field> keyValueChildren = keyValueStructField.getChildren();
+        Field keyField;
+        Field valueField;
+        if (keyValueChildren.size() != 2) {
+            throw new IllegalStateException("Invalid Arrow Map schema: " + field);
+        }
+        else {
+            keyField = keyValueChildren.get(0);
+            valueField = keyValueChildren.get(1);
+            if (!KEY.equals(keyField.getName()) || !VALUE.equals(valueField.getName())) {
+                throw new IllegalStateException("Invalid Arrow Map schema: " + field);
+            }
+        }
+
+        for (Field nextChild : keyValueChildren) {
+            //For each child field that comprises the Map, attempt to extract and write the corresponding value
+            //using the FieldResolver.
+            Object childValue = resolver.getFieldValue(nextChild, value);
+            switch (Types.getMinorTypeForArrowType(nextChild.getType())) {
+                case LIST:
+                    writeList(allocator,
+                            (FieldWriter) writer.list(nextChild.getName()),
+                            nextChild,
+                            pos,
+                            ((List) childValue),
+                            resolver);
+                    break;
+                case STRUCT:
+                    writeStruct(allocator,
+                            writer.struct(nextChild.getName()),
+                            nextChild,
+                            pos,
+                            childValue,
+                            resolver);
+                    break;
+                default:
+                    writeMapValue(writer, nextChild, allocator, childValue);
+                    break;
+            }
+        }
+        writer.endEntry();
+    }
+
+    /**
+     * Used to write a value into a specific child field within a Map. Multiple calls to this method per-cell are
+     * expected in order to write to all N fields of a Map.
+     *
+     * @param writer The FieldWriter (already positioned at the row and list entry number) that we want to write into.
+     * @param field The child field we are attempting to write into.
+     * @param allocator The BlockAllocator that can be used for allocating Arrow Buffers for fields which require conversion
+     * to Arrow Buff before being written.
+     * @param value The value to write.
+     * @note This method and its List complement violate the DRY mantra because ListWriter and StructWriter don't share
```

Comment:
do you mean ListWriter and _UnionMapWriter_?

Reply:
oops, typo.